File: app/src/main/java/org/secuso/privacyfriendlydicer/MainActivity.java
package org.secuso.privacyfriendlydicer;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.os.Vibrator;
import android.preference.PreferenceManager;

import com.google.android.material.navigation.NavigationView;

import androidx.appcompat.widget.DrawableUtils;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.view.MenuItem;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    private static Dicer dicer = new Dicer();

    private ImageView[] imageViews;
    private boolean shakingEnabled;
    private boolean vibrationEnabled;
    private SharedPreferences sharedPreferences;

    // for Shaking
    private SensorManager sensorManager;
    private Sensor accelerometer;
    private ShakeListener shakeListener;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.setDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = (NavigationView) findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        //Preferences
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getBaseContext());

        //Seekbars
        final SeekBar poolSeekBar = (SeekBar) findViewById(R.id.seekBar);

        poolSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                TextView textViewLengthDisplay =
                        (TextView) findViewById(R.id.chooseDiceNumber);
                textViewLengthDisplay.setText(Integer.toString(progress + 1));
            }

            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        final SeekBar facesSeekBar = (SeekBar) findViewById(R.id.seekBarFace);

        facesSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                TextView textViewLengthDisplay =
                        (TextView) findViewById(R.id.chooseFaceNumber);
                textViewLengthDisplay.setText(Integer.toString(progress + 1));
            }

            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        //Button
        Button rollDiceButton = (Button) findViewById(R.id.rollButton);

        rollDiceButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                evaluate((Vibrator) getSystemService(Context.VIBRATOR_SERVICE), poolSeekBar.getProgress() + 1, facesSeekBar.getProgress() + 1);

            }
        });

        //Shaking
        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        accelerometer = sensorManager
                .getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        shakeListener = new ShakeListener();
        shakeListener.setOnShakeListener(new ShakeListener.OnShakeListener() {

            public void onShake(int count) {

                if (shakingEnabled) {
                    evaluate((Vibrator) getSystemService(Context.VIBRATOR_SERVICE), poolSeekBar.getProgress() + 1, facesSeekBar.getProgress() + 1);
                }
            }
        });

        displaySum(new int[]{0});
        initResultDiceViews();
    }

    public void flashResult(ImageView imageView) {

        Animation animation = new AlphaAnimation(0.0f, 1.0f);
        animation.setDuration(500);
        animation.setStartOffset(20);
        animation.setRepeatMode(Animation.REVERSE);
        imageView.startAnimation(animation);

    }

    public void initResultDiceViews() {
        imageViews = new ImageView[10];

        imageViews[0] = (ImageView) findViewById(R.id.resultOne);
        imageViews[1] = (ImageView) findViewById(R.id.resultTwo);
        imageViews[2] = (ImageView) findViewById(R.id.resultThree);
        imageViews[3] = (ImageView) findViewById(R.id.resultFour);
        imageViews[4] = (ImageView) findViewById(R.id.resultFive);
        imageViews[5] = (ImageView) findViewById(R.id.resultSix);
        imageViews[6] = (ImageView) findViewById(R.id.resultSeven);
        imageViews[7] = (ImageView) findViewById(R.id.resultEight);
        imageViews[8] = (ImageView) findViewById(R.id.resultNine);
        imageViews[9] = (ImageView) findViewById(R.id.resultTen);

        for (int i = 0; i < imageViews.length; i++) {
            imageViews[i].setImageResource(0);
        }
    }

    public void switchDice(ImageView imageView, int result) {

        imageView.setImageResource(this.getResources().getIdentifier("d"+result, "drawable", this.getPackageName()));
    }

    public void evaluate(Vibrator vibrator, int diceNumber, int faceNumber) {

        applySettings();

        int[] dice = dicer.rollDice(diceNumber, faceNumber);

        displaySum(dice);
        initResultDiceViews();
        showDice(dice);

        if (vibrationEnabled) {
            vibrator.vibrate(50);
        }
    }

    private void showDice(int[] dice) {
        for (int i = 0; i < dice.length; i++) {
            switchDice(imageViews[i], dice[i]);
            flashResult(imageViews[i]);
        }
    }

    private void displaySum(int[] dice) {
        int sum = 0;
        for (int d : dice) {
            sum += d;
        }
        TextView sumTextView = findViewById(R.id.sumTextView);
        sumTextView.setText(getString(R.string.main_dice_sum, Integer.toString(sum)));
    }

    public void applySettings() {
        shakingEnabled = sharedPreferences.getBoolean("enable_shaking", true);
        vibrationEnabled = sharedPreferences.getBoolean("enable_vibration", true);
    }

    @Override
    public void onResume() {
        super.onResume();
        sensorManager.registerListener(shakeListener, accelerometer,
                SensorManager.SENSOR_DELAY_UI);

        applySettings();

    }

    @Override
    public void onPause() {
        sensorManager.unregisterListener(shakeListener);
        super.onPause();
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        Intent intent;

        switch (item.getItemId()) {
            case R.id.nav_about:
                intent = new Intent(this, AboutActivity.class);
                startActivity(intent);
                return true;

            case R.id.nav_help:
                intent = new Intent(this, HelpActivity.class);
                startActivity(intent);
                return true;

            case R.id.nav_settimgs:
                intent = new Intent(this, SettingsActivity.class);
                startActivity(intent);
                return true;

            case R.id.nav_tutorial:
                intent = new Intent(this, TutorialActivity.class);
                startActivity(intent);
                return true;

            default:
        }

        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }
}

File: app/src/main/java/org/secuso/privacyfriendlydicer/ui/MainActivity.kt
package org.secuso.privacyfriendlydicer.ui

import android.content.Intent
import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.Bundle
import android.os.Vibrator
import android.preference.PreferenceManager
import android.view.MenuItem
import android.view.View
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import com.google.android.material.navigation.NavigationView
import org.secuso.privacyfriendlydicer.checkGoodbyeGoogle
import org.secuso.privacyfriendlydicer.R
import org.secuso.privacyfriendlydicer.databinding.ActivityMainBinding
import org.secuso.privacyfriendlydicer.databinding.ContentMainBinding
import org.secuso.privacyfriendlydicer.sensors.ShakeListener
import org.secuso.privacyfriendlydicer.sensors.ShakeListener.OnShakeListener
import java.util.Locale

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    private val dicerViewModel by lazy { ViewModelProvider(this).get<DicerViewModel>(DicerViewModel::class.java) }

    private val binding by lazy { ActivityMainBinding.inflate(layoutInflater) }
    private val contentMainBinding by lazy { ContentMainBinding.bind(binding.root) }

    private val adapter: DiceAdapter by lazy { DiceAdapter(dicerViewModel.dicerLiveData.value ?: IntArray(0), layoutInflater) }
    private var shakingEnabled = false
    private var vibrationEnabled = false
    private val sharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(this) }

    // for Shaking
    private val sensorManager: SensorManager by lazy { getSystemService(SENSOR_SERVICE) as SensorManager }
    private val accelerometer: Sensor? by lazy { sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) }
    private val shakeListener = ShakeListener()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        
        contentMainBinding.dices.adapter = adapter
        contentMainBinding.dices.layoutManager = object : GridLayoutManager(this, 5, VERTICAL, false) {
            override fun canScrollHorizontally() = false
            override fun canScrollVertically() = false
        }

        initResources()

        dicerViewModel.dicerLiveData.observe(this, object : Observer<IntArray> {
            override fun onChanged(dice: IntArray) {
                adapter.dices = dice
                displaySum()

                if (vibrationEnabled) {
                    val vibrator = getSystemService(VIBRATOR_SERVICE) as Vibrator
                    vibrator.vibrate(50)
                }
            }
        })

        dicerViewModel.diceNumberLiveData.observe(this, object : Observer<Int> {
            override fun onChanged(number: Int) {
                contentMainBinding.chooseDiceNumber.text = String.format(Locale.ENGLISH, "%d", number)
            }
        })

        dicerViewModel.faceNumberLiveData.observe(this, object : Observer<Int> {
            override fun onChanged(number: Int) {
                contentMainBinding.chooseFaceNumber.text = String.format(Locale.ENGLISH, "%d", number)
            }
        })
        checkGoodbyeGoogle(this, layoutInflater)
    }

    private fun initResources() {
        setSupportActionBar(binding.toolbar)

        val toggle = ActionBarDrawerToggle(this, binding.drawerLayout, binding.toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        binding.drawerLayout.addDrawerListener(toggle)
        toggle.syncState()

        binding.navView.setNavigationItemSelectedListener(this)

        contentMainBinding.seekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                dicerViewModel.setDiceNumber(progress + 1)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
            }
        })

        contentMainBinding.seekBarFace.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                dicerViewModel.setFaceNumber(progress + 1)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
            }
        })

        //Button
        contentMainBinding.rollButton.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                rollDice()
            }
        })

        //Shaking
        shakeListener.setOnShakeListener(object : OnShakeListener {
            override fun onShake(count: Int) {
                if (shakingEnabled) {
                    rollDice()
                }
            }
        })
    }

    fun rollDice() {
        applySettings()
        dicerViewModel.rollDice()
    }

    private fun displaySum() {
        contentMainBinding.sumTextView.text = getString(R.string.main_dice_sum, adapter.dices.sum().toString())
    }

    fun applySettings() {
        shakingEnabled = sharedPreferences.getBoolean("enable_shaking", true)
        vibrationEnabled = sharedPreferences.getBoolean("enable_vibration", true)
    }

    public override fun onResume() {
        super.onResume()
        sensorManager.registerListener(
            shakeListener, accelerometer,
            SensorManager.SENSOR_DELAY_UI
        )

        applySettings()
    }

    public override fun onPause() {
        sensorManager.unregisterListener(shakeListener)
        super.onPause()
    }

    override fun onBackPressed() {
        if (binding.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            binding.drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }


    /**
     * Handle navigation view item clicks here.
     */
    override fun onNavigationItemSelected(item: MenuItem): Boolean {

        val intent = when (item.itemId) {
            R.id.nav_about -> Intent(this, AboutActivity::class.java)
            R.id.nav_help -> Intent(this, HelpActivity::class.java)
            R.id.nav_settimgs -> Intent(this, SettingsActivity::class.java)
            R.id.nav_tutorial -> Intent(this, TutorialActivity::class.java)
            else -> return false
        }

        startActivity(intent)
        binding.drawerLayout.closeDrawer(GravityCompat.START)
        return true
    }
}
